File: Kernel Simulator/Network/Transfer/NetworkTransfer.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.IO
Imports System.Net.Http
Imports System.Net.Http.Handlers
Imports System.Threading
Imports System.Threading.Tasks
Imports KS.Files.Folders
Imports KS.Misc.Notifications

Namespace Network.Transfer
    Public Module NetworkTransfer

        Public DownloadPercentagePrint As String = ""
        Public UploadPercentagePrint As String = ""
        Public DownloadNotificationProvoke As Boolean
        Public UploadNotificationProvoke As Boolean
        Friend IsError As Boolean
        Friend ReasonError As Exception
        Friend CancellationToken As New CancellationTokenSource
        Friend WClientProgress As New ProgressMessageHandler(New HttpClientHandler)
        Friend WClient As New HttpClient(WClientProgress)
        Friend DownloadedString As String
        Friend DownloadNotif As Notification
        Friend UploadNotif As Notification
        Friend SuppressDownloadMessage As Boolean
        Friend SuppressUploadMessage As Boolean

        ''' <summary>
        ''' Downloads a file to the current working directory.
        ''' </summary>
        ''' <param name="URL">A URL to a file</param>
        ''' <returns>True if successful. Throws exception if unsuccessful.</returns>
        Public Function DownloadFile(URL As String) As Boolean
            Return DownloadFile(URL, ShowProgress)
        End Function

        ''' <summary>
        ''' Downloads a file to the current working directory.
        ''' </summary>
        ''' <param name="URL">A URL to a file</param>
        ''' <param name="ShowProgress">Whether or not to show progress bar</param>
        ''' <returns>True if successful. Throws exception if unsuccessful.</returns>
        Public Function DownloadFile(URL As String, ShowProgress As Boolean) As Boolean
            Dim FileName As String = GetFilenameFromUrl(URL)
            Return DownloadFile(URL, ShowProgress, FileName)
        End Function

        ''' <summary>
        ''' Downloads a file to the current working directory.
        ''' </summary>
        ''' <param name="URL">A URL to a file</param>
        ''' <param name="FileName">File name to download to</param>
        ''' <returns>True if successful. Throws exception if unsuccessful.</returns>
        Public Function DownloadFile(URL As String, FileName As String) As Boolean
            Return DownloadFile(URL, ShowProgress, FileName)
        End Function

        ''' <summary>
        ''' Downloads a file to the current working directory.
        ''' </summary>
        ''' <param name="URL">A URL to a file</param>
        ''' <param name="ShowProgress">Whether or not to show progress bar</param>
        ''' <param name="FileName">File name to download to</param>
        ''' <returns>True if successful. Throws exception if unsuccessful.</returns>
        Public Function DownloadFile(URL As String, ShowProgress As Boolean, FileName As String) As Boolean
            'Intialize variables
            Dim FileUri As New Uri(URL)

            'Initialize the progress bar indicator and the file completed event handler
            If DownloadNotificationProvoke Then
                DownloadNotif = New Notification(DoTranslation("Downloading..."), FileUri.AbsoluteUri, NotifPriority.Low, NotifType.Progress)
                NotifySend(DownloadNotif)
            End If
            If ShowProgress Then AddHandler WClientProgress.HttpReceiveProgress, AddressOf HttpReceiveProgressWatch

            'Send the GET request to the server for the file
            Wdbg(DebugLevel.I, "Directory location: {0}", CurrentDir)
            Dim Response As HttpResponseMessage = WClient.GetAsync(FileUri, CancellationToken.Token).Result
            Response.EnsureSuccessStatusCode()

            'Get the file stream
            Dim FilePath As String = NeutralizePath(FileName)
            Dim FileStream As New FileStream(FilePath, FileMode.Create, FileAccess.Write)

            'Try to download the file asynchronously
            Task.Run(Sub()
                         Try
                             Response.Content.ReadAsStreamAsync.Result.CopyTo(FileStream)
                             FileStream.Flush()
                             FileStream.Close()
                             DownloadChecker(Nothing)
                         Catch ex As Exception
                             DownloadChecker(ex)
                         End Try
                     End Sub, CancellationToken.Token)
            While Not TransferFinished
                If CancelRequested Then
                    TransferFinished = True
                    CancellationToken.Cancel()
                End If
            End While

            'We're done downloading. Check to see if it's actually an error
            TransferFinished = False
            If ShowProgress And Not SuppressDownloadMessage Then WritePlain("", True)
            SuppressDownloadMessage = False
            If IsError Then
                If DownloadNotificationProvoke Then DownloadNotif.ProgressFailed = True
                CancellationToken.Cancel()
                Throw ReasonError
            Else
                If DownloadNotificationProvoke Then DownloadNotif.Progress = 100
                Return True
            End If
        End Function

        ''' <summary>
        ''' Uploads a file to the current working directory.
        ''' </summary>
        ''' <param name="FileName">A target file name. Use <see cref="NeutralizePath(String, Boolean)"/> to get full path of source.</param>
        ''' <param name="URL">A URL to a file</param>
        ''' <returns>True if successful. Throws exception if unsuccessful.</returns>
        Public Function UploadFile(FileName As String, URL As String) As Boolean
            Return UploadFile(FileName, URL, ShowProgress)
        End Function

        ''' <summary>
        ''' Uploads a file from the current working directory.
        ''' </summary>
        ''' <param name="FileName">A target file name. Use <see cref="NeutralizePath(String, Boolean)"/> to get full path of source.</param>
        ''' <param name="URL">A URL</param>
        ''' <param name="ShowProgress">Whether or not to show progress bar</param>
        ''' <returns>True if successful. Throws exception if unsuccessful.</returns>
        Public Function UploadFile(FileName As String, URL As String, ShowProgress As Boolean) As Boolean
            'Intialize variables
            Dim FileUri As New Uri(URL)

            'Initialize the progress bar indicator and the file completed event handler
            If UploadNotificationProvoke Then
                UploadNotif = New Notification(DoTranslation("Uploading..."), FileUri.AbsoluteUri, NotifPriority.Low, NotifType.Progress)
                NotifySend(DownloadNotif)
            End If
            If ShowProgress Then AddHandler WClientProgress.HttpSendProgress, AddressOf HttpSendProgressWatch

            'Send the GET request to the server for the file after getting the stream and target file stream
            Wdbg(DebugLevel.I, "Directory location: {0}", CurrentDir)
            Dim FilePath As String = NeutralizePath(FileName)
            Dim FileStream As New FileStream(FilePath, FileMode.Open, FileAccess.Read)
            Dim Content As New StreamContent(FileStream)

            'Upload now
            Try
                Dim Response As HttpResponseMessage = WClient.PutAsync(URL, Content, CancellationToken.Token).Result
                Response.EnsureSuccessStatusCode()
                UploadChecker(Nothing)
            Catch ex As Exception
                UploadChecker(ex)
            End Try

            'We're done uploading. Check to see if it's actually an error
            TransferFinished = False
            If ShowProgress And Not SuppressUploadMessage Then WritePlain("", True)
            SuppressUploadMessage = False
            If IsError Then
                If UploadNotificationProvoke Then UploadNotif.ProgressFailed = True
                CancellationToken.Cancel()
                Throw ReasonError
            Else
                If UploadNotificationProvoke Then UploadNotif.Progress = 100
                Return True
            End If
        End Function

        ''' <summary>
        ''' Downloads a resource from URL as a string.
        ''' </summary>
        ''' <param name="URL">A URL to a file</param>
        ''' <returns>True if successful. Throws exception if unsuccessful.</returns>
        Public Function DownloadString(URL As String) As String
            Return DownloadString(URL, ShowProgress)
        End Function

        ''' <summary>
        ''' Downloads a resource from URL as a string.
        ''' </summary>
        ''' <param name="URL">A URL</param>
        ''' <param name="ShowProgress">Whether or not to show progress bar</param>
        ''' <returns>A resource string if successful; Throws exception if unsuccessful.</returns>
        Public Function DownloadString(URL As String, ShowProgress As Boolean) As String
            'Intialize variables
            Dim StringUri As New Uri(URL)

            'Initialize the progress bar indicator and the file completed event handler
            If DownloadNotificationProvoke Then
                DownloadNotif = New Notification(DoTranslation("Downloading..."), StringUri.AbsoluteUri, NotifPriority.Low, NotifType.Progress)
                NotifySend(DownloadNotif)
            End If
            If ShowProgress Then AddHandler WClientProgress.HttpReceiveProgress, AddressOf HttpReceiveProgressWatch

            'Send the GET request to the server for the file
            Wdbg(DebugLevel.I, "Directory location: {0}", CurrentDir)
            Dim Response As HttpResponseMessage = WClient.GetAsync(StringUri, CancellationToken.Token).Result
            Response.EnsureSuccessStatusCode()

            'Get the memory stream
            Dim ContentStream As New MemoryStream

            'Try to download the string asynchronously
            Task.Run(Sub()
                         Try
                             Response.Content.ReadAsStreamAsync.Result.CopyTo(ContentStream)
                             ContentStream.Seek(0, SeekOrigin.Begin)
                             DownloadChecker(Nothing)
                         Catch ex As Exception
                             DownloadChecker(ex)
                         End Try
                     End Sub, CancellationToken.Token)
            While Not TransferFinished
                If CancelRequested Then
                    TransferFinished = True
                    CancellationToken.Cancel()
                End If
            End While

            'We're done downloading. Check to see if it's actually an error
            TransferFinished = False
            If ShowProgress And Not SuppressDownloadMessage Then WritePlain("", True)
            SuppressDownloadMessage = False
            If IsError Then
                If DownloadNotificationProvoke Then DownloadNotif.ProgressFailed = True
                CancellationToken.Cancel()
                Throw ReasonError
            Else
                If DownloadNotificationProvoke Then DownloadNotif.Progress = 100
                Return New StreamReader(ContentStream).ReadToEnd
            End If
        End Function

        ''' <summary>
        ''' Uploads a resource from URL as a string.
        ''' </summary>
        ''' <param name="URL">A URL to a file</param>
        ''' <param name="Data">Content to upload</param>
        ''' <returns>True if successful. Throws exception if unsuccessful.</returns>
        Public Function UploadString(URL As String, Data As String) As Boolean
            Return UploadString(URL, Data, ShowProgress)
        End Function

        ''' <summary>
        ''' Uploads a resource from URL as a string.
        ''' </summary>
        ''' <param name="URL">A URL</param>
        ''' <param name="Data">Content to upload</param>
        ''' <param name="ShowProgress">Whether or not to show progress bar</param>
        ''' <returns>A resource string if successful; Throws exception if unsuccessful.</returns>
        Public Function UploadString(URL As String, Data As String, ShowProgress As Boolean) As Boolean
            'Intialize variables
            Dim StringUri As New Uri(URL)

            'Initialize the progress bar indicator and the file completed event handler
            If UploadNotificationProvoke Then
                UploadNotif = New Notification(DoTranslation("Uploading..."), StringUri.AbsoluteUri, NotifPriority.Low, NotifType.Progress)
                NotifySend(UploadNotif)
            End If
            If ShowProgress Then AddHandler WClientProgress.HttpSendProgress, AddressOf HttpSendProgressWatch

            'Send the GET request to the server for the file
            Wdbg(DebugLevel.I, "Directory location: {0}", CurrentDir)
            Dim StringContent As New StringContent(Data)

            Try
                Dim Response As HttpResponseMessage = WClient.PutAsync(URL, StringContent, CancellationToken.Token).Result
                Response.EnsureSuccessStatusCode()
                UploadChecker(Nothing)
            Catch ex As Exception
                UploadChecker(ex)
            End Try

            'We're done uploading. Check to see if it's actually an error
            TransferFinished = False
            If ShowProgress And Not SuppressUploadMessage Then WritePlain("", True)
            SuppressUploadMessage = False
            If IsError Then
                If UploadNotificationProvoke Then UploadNotif.ProgressFailed = True
                CancellationToken.Cancel()
                Throw ReasonError
            Else
                If UploadNotificationProvoke Then UploadNotif.Progress = 100
                Return True
            End If
        End Function

        ''' <summary>
        ''' Check for errors on download completion.
        ''' </summary>
        Private Sub DownloadChecker(e As Exception)
            Wdbg(DebugLevel.I, "Download complete. Error: {0}", e?.Message)
            If e IsNot Nothing Then
                If DownloadNotificationProvoke Then DownloadNotif.ProgressFailed = True
                ReasonError = e
                IsError = True
            End If
            TransferFinished = True
        End Sub

        ''' <summary>
        ''' Thread to check for errors on download completion.
        ''' </summary>
        Private Sub UploadChecker(e As Exception)
            Wdbg(DebugLevel.I, "Upload complete. Error: {0}", e?.Message)
            If e IsNot Nothing Then
                If UploadNotificationProvoke Then UploadNotif.ProgressFailed = True
                ReasonError = e
                IsError = True
            End If
            TransferFinished = True
        End Sub

        Private Sub HttpReceiveProgressWatch(sender As Object, e As HttpProgressEventArgs)
            Dim TotalBytes As Integer = If(e.TotalBytes, -1)
            Dim TransferInfo As New NetworkTransferInfo(e.BytesTransferred, TotalBytes, NetworkTransferType.Download)
            SuppressDownloadMessage = TotalBytes = -1
            TransferProgress(TransferInfo)
        End Sub

        Private Sub HttpSendProgressWatch(sender As Object, e As HttpProgressEventArgs)
            Dim TotalBytes As Integer = If(e.TotalBytes, -1)
            Dim TransferInfo As New NetworkTransferInfo(e.BytesTransferred, TotalBytes, NetworkTransferType.Upload)
            SuppressUploadMessage = TotalBytes = -1
            TransferProgress(TransferInfo)
        End Sub

        ''' <summary>
        ''' Report the progress to the console.
        ''' </summary>
        Private Sub TransferProgress(TransferInfo As NetworkTransferInfo)
            Try
                'Distinguish download from upload
                Dim NotificationProvoke As Boolean = If(TransferInfo.TransferType = NetworkTransferType.Download, DownloadNotificationProvoke, UploadNotificationProvoke)
                Dim NotificationInstance As Notification = If(TransferInfo.TransferType = NetworkTransferType.Download, DownloadNotif, UploadNotif)

                'Report the progress
                If Not TransferFinished Then
                    If TransferInfo.FileSize >= 0 And Not TransferInfo.MessageSuppressed Then
                        'We know the total bytes. Print it out.
                        Dim Progress As Double = 100 * (TransferInfo.DoneSize / TransferInfo.FileSize)
                        If NotificationProvoke Then
                            NotificationInstance.Progress = Progress
                        Else
                            If Not String.IsNullOrWhiteSpace(DownloadPercentagePrint) Then
                                WriteWhere(ProbePlaces(DownloadPercentagePrint), 0, ConsoleWrapper.CursorTop, False, GetConsoleColor(ColTypes.Neutral), TransferInfo.DoneSize.FileSizeToString, TransferInfo.FileSize.FileSizeToString, Progress)
                            Else
                                WriteWhere(DoTranslation("{0} of {1} downloaded.") + " | {2}%", 0, ConsoleWrapper.CursorTop, False, GetConsoleColor(ColTypes.Neutral), TransferInfo.DoneSize.FileSizeToString, TransferInfo.FileSize.FileSizeToString, Progress)
                            End If
                            ClearLineToRight()
                        End If
                    Else
                        TransferInfo.MessageSuppressed = True
                    End If
                End If
            Catch ex As Exception
                Wdbg(DebugLevel.E, "Error trying to report transfer progress: {0}", ex.Message)
                WStkTrc(ex)
            End Try
        End Sub

    End Module
End Namespace

File: Kernel Simulator/Network/Transfer/NetworkTransfer.cs
//
// Kernel Simulator  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Kernel Simulator
//
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System.IO;
using System.Net.Http;
using System.Net.Http.Handlers;
using System.Threading;
using System.Threading.Tasks;
using FluentFTP.Helpers;
using KS.ConsoleBase.Colors;
using KS.Files;
using KS.Files.Folders;
using KS.Kernel;
using KS.Languages;
using KS.Misc.Notifiers;
using KS.Misc.Probers;
using KS.Misc.Writers.ConsoleWriters;
using KS.Misc.Writers.DebugWriters;
using Terminaux.Base;

namespace KS.Network.Transfer
{
    public static class NetworkTransfer
    {

        public static string DownloadPercentagePrint = "";
        public static string UploadPercentagePrint = "";
        public static bool DownloadNotificationProvoke;
        public static bool UploadNotificationProvoke;
        internal static bool IsError;
        internal static Exception ReasonError;
        internal static CancellationTokenSource CancellationToken = new();
        internal static ProgressMessageHandler WClientProgress = new(new HttpClientHandler());
        internal static HttpClient WClient = new(WClientProgress);
        internal static string DownloadedString;
        internal static Notification DownloadNotif;
        internal static Notification UploadNotif;
        internal static bool SuppressDownloadMessage;
        internal static bool SuppressUploadMessage;

        /// <summary>
        /// Downloads a file to the current working directory.
        /// </summary>
        /// <param name="URL">A URL to a file</param>
        /// <returns>True if successful. Throws exception if unsuccessful.</returns>
        public static bool DownloadFile(string URL)
        {
            return DownloadFile(URL, Flags.ShowProgress);
        }

        /// <summary>
        /// Downloads a file to the current working directory.
        /// </summary>
        /// <param name="URL">A URL to a file</param>
        /// <param name="ShowProgress">Whether or not to show progress bar</param>
        /// <returns>True if successful. Throws exception if unsuccessful.</returns>
        public static bool DownloadFile(string URL, bool ShowProgress)
        {
            string FileName = NetworkTools.GetFilenameFromUrl(URL);
            return DownloadFile(URL, ShowProgress, FileName);
        }

        /// <summary>
        /// Downloads a file to the current working directory.
        /// </summary>
        /// <param name="URL">A URL to a file</param>
        /// <param name="FileName">File name to download to</param>
        /// <returns>True if successful. Throws exception if unsuccessful.</returns>
        public static bool DownloadFile(string URL, string FileName)
        {
            return DownloadFile(URL, Flags.ShowProgress, FileName);
        }

        /// <summary>
        /// Downloads a file to the current working directory.
        /// </summary>
        /// <param name="URL">A URL to a file</param>
        /// <param name="ShowProgress">Whether or not to show progress bar</param>
        /// <param name="FileName">File name to download to</param>
        /// <returns>True if successful. Throws exception if unsuccessful.</returns>
        public static bool DownloadFile(string URL, bool ShowProgress, string FileName)
        {
            // Intialize variables
            var FileUri = new Uri(URL);

            // Initialize the progress bar indicator and the file completed event handler
            if (DownloadNotificationProvoke)
            {
                DownloadNotif = new Notification(Translate.DoTranslation("Downloading..."), FileUri.AbsoluteUri, Notifications.NotifPriority.Low, Notifications.NotifType.Progress);
                Notifications.NotifySend(DownloadNotif);
            }
            if (ShowProgress)
                WClientProgress.HttpReceiveProgress += HttpReceiveProgressWatch;

            // Send the GET request to the server for the file
            DebugWriter.Wdbg(DebugLevel.I, "Directory location: {0}", CurrentDirectory.CurrentDir);
            var Response = WClient.GetAsync(FileUri, CancellationToken.Token).Result;
            Response.EnsureSuccessStatusCode();

            // Get the file stream
            string FilePath = Filesystem.NeutralizePath(FileName);
            var FileStream = new FileStream(FilePath, FileMode.Create, FileAccess.Write);

            // Try to download the file asynchronously
            Task.Run(() => { try { Response.Content.ReadAsStreamAsync().Result.CopyTo(FileStream); FileStream.Flush(); FileStream.Close(); DownloadChecker(null); } catch (Exception ex) { DownloadChecker(ex); } }, CancellationToken.Token);
            while (!NetworkTools.TransferFinished)
            {
                if (Flags.CancelRequested)
                {
                    NetworkTools.TransferFinished = true;
                    CancellationToken.Cancel();
                }
            }

            // We're done downloading. Check to see if it's actually an error
            NetworkTools.TransferFinished = false;
            if (ShowProgress & !SuppressDownloadMessage)
                TextWriterColor.WritePlain("", true);
            SuppressDownloadMessage = false;
            if (IsError)
            {
                if (DownloadNotificationProvoke)
                    DownloadNotif.ProgressFailed = true;
                CancellationToken.Cancel();
                throw ReasonError;
            }
            else
            {
                if (DownloadNotificationProvoke)
                    DownloadNotif.Progress = 100;
                return true;
            }
        }

        /// <summary>
        /// Uploads a file to the current working directory.
        /// </summary>
        /// <param name="FileName">A target file name. Use <see cref="Filesystem.NeutralizePath(string, bool)"/> to get full path of source.</param>
        /// <param name="URL">A URL to a file</param>
        /// <returns>True if successful. Throws exception if unsuccessful.</returns>
        public static bool UploadFile(string FileName, string URL)
        {
            return UploadFile(FileName, URL, Flags.ShowProgress);
        }

        /// <summary>
        /// Uploads a file from the current working directory.
        /// </summary>
        /// <param name="FileName">A target file name. Use <see cref="Filesystem.NeutralizePath(string, bool)"/> to get full path of source.</param>
        /// <param name="URL">A URL</param>
        /// <param name="ShowProgress">Whether or not to show progress bar</param>
        /// <returns>True if successful. Throws exception if unsuccessful.</returns>
        public static bool UploadFile(string FileName, string URL, bool ShowProgress)
        {
            // Intialize variables
            var FileUri = new Uri(URL);

            // Initialize the progress bar indicator and the file completed event handler
            if (UploadNotificationProvoke)
            {
                UploadNotif = new Notification(Translate.DoTranslation("Uploading..."), FileUri.AbsoluteUri, Notifications.NotifPriority.Low, Notifications.NotifType.Progress);
                Notifications.NotifySend(DownloadNotif);
            }
            if (ShowProgress)
                WClientProgress.HttpSendProgress += HttpSendProgressWatch;

            // Send the GET request to the server for the file after getting the stream and target file stream
            DebugWriter.Wdbg(DebugLevel.I, "Directory location: {0}", CurrentDirectory.CurrentDir);
            string FilePath = Filesystem.NeutralizePath(FileName);
            var FileStream = new FileStream(FilePath, FileMode.Open, FileAccess.Read);
            var Content = new StreamContent(FileStream);

            // Upload now
            try
            {
                var Response = WClient.PutAsync(URL, Content, CancellationToken.Token).Result;
                Response.EnsureSuccessStatusCode();
                UploadChecker(null);
            }
            catch (Exception ex)
            {
                UploadChecker(ex);
            }

            // We're done uploading. Check to see if it's actually an error
            NetworkTools.TransferFinished = false;
            if (ShowProgress & !SuppressUploadMessage)
                TextWriterColor.WritePlain("", true);
            SuppressUploadMessage = false;
            if (IsError)
            {
                if (UploadNotificationProvoke)
                    UploadNotif.ProgressFailed = true;
                CancellationToken.Cancel();
                throw ReasonError;
            }
            else
            {
                if (UploadNotificationProvoke)
                    UploadNotif.Progress = 100;
                return true;
            }
        }

        /// <summary>
        /// Downloads a resource from URL as a string.
        /// </summary>
        /// <param name="URL">A URL to a file</param>
        /// <returns>True if successful. Throws exception if unsuccessful.</returns>
        public static string DownloadString(string URL)
        {
            return DownloadString(URL, Flags.ShowProgress);
        }

        /// <summary>
        /// Downloads a resource from URL as a string.
        /// </summary>
        /// <param name="URL">A URL</param>
        /// <param name="ShowProgress">Whether or not to show progress bar</param>
        /// <returns>A resource string if successful; Throws exception if unsuccessful.</returns>
        public static string DownloadString(string URL, bool ShowProgress)
        {
            // Intialize variables
            var StringUri = new Uri(URL);

            // Initialize the progress bar indicator and the file completed event handler
            if (DownloadNotificationProvoke)
            {
                DownloadNotif = new Notification(Translate.DoTranslation("Downloading..."), StringUri.AbsoluteUri, Notifications.NotifPriority.Low, Notifications.NotifType.Progress);
                Notifications.NotifySend(DownloadNotif);
            }
            if (ShowProgress)
                WClientProgress.HttpReceiveProgress += HttpReceiveProgressWatch;

            // Send the GET request to the server for the file
            DebugWriter.Wdbg(DebugLevel.I, "Directory location: {0}", CurrentDirectory.CurrentDir);
            var Response = WClient.GetAsync(StringUri, CancellationToken.Token).Result;
            Response.EnsureSuccessStatusCode();

            // Get the memory stream
            var ContentStream = new MemoryStream();

            // Try to download the string asynchronously
            Task.Run(() => { try { Response.Content.ReadAsStreamAsync().Result.CopyTo(ContentStream); ContentStream.Seek(0L, SeekOrigin.Begin); DownloadChecker(null); } catch (Exception ex) { DownloadChecker(ex); } }, CancellationToken.Token);
            while (!NetworkTools.TransferFinished)
            {
                if (Flags.CancelRequested)
                {
                    NetworkTools.TransferFinished = true;
                    CancellationToken.Cancel();
                }
            }

            // We're done downloading. Check to see if it's actually an error
            NetworkTools.TransferFinished = false;
            if (ShowProgress & !SuppressDownloadMessage)
                TextWriterColor.WritePlain("", true);
            SuppressDownloadMessage = false;
            if (IsError)
            {
                if (DownloadNotificationProvoke)
                    DownloadNotif.ProgressFailed = true;
                CancellationToken.Cancel();
                throw ReasonError;
            }
            else
            {
                if (DownloadNotificationProvoke)
                    DownloadNotif.Progress = 100;
                return new StreamReader(ContentStream).ReadToEnd();
            }
        }

        /// <summary>
        /// Uploads a resource from URL as a string.
        /// </summary>
        /// <param name="URL">A URL to a file</param>
        /// <param name="Data">Content to upload</param>
        /// <returns>True if successful. Throws exception if unsuccessful.</returns>
        public static bool UploadString(string URL, string Data)
        {
            return UploadString(URL, Data, Flags.ShowProgress);
        }

        /// <summary>
        /// Uploads a resource from URL as a string.
        /// </summary>
        /// <param name="URL">A URL</param>
        /// <param name="Data">Content to upload</param>
        /// <param name="ShowProgress">Whether or not to show progress bar</param>
        /// <returns>A resource string if successful; Throws exception if unsuccessful.</returns>
        public static bool UploadString(string URL, string Data, bool ShowProgress)
        {
            // Intialize variables
            var StringUri = new Uri(URL);

            // Initialize the progress bar indicator and the file completed event handler
            if (UploadNotificationProvoke)
            {
                UploadNotif = new Notification(Translate.DoTranslation("Uploading..."), StringUri.AbsoluteUri, Notifications.NotifPriority.Low, Notifications.NotifType.Progress);
                Notifications.NotifySend(UploadNotif);
            }
            if (ShowProgress)
                WClientProgress.HttpSendProgress += HttpSendProgressWatch;

            // Send the GET request to the server for the file
            DebugWriter.Wdbg(DebugLevel.I, "Directory location: {0}", CurrentDirectory.CurrentDir);
            var StringContent = new StringContent(Data);

            try
            {
                var Response = WClient.PutAsync(URL, StringContent, CancellationToken.Token).Result;
                Response.EnsureSuccessStatusCode();
                UploadChecker(null);
            }
            catch (Exception ex)
            {
                UploadChecker(ex);
            }

            // We're done uploading. Check to see if it's actually an error
            NetworkTools.TransferFinished = false;
            if (ShowProgress & !SuppressUploadMessage)
                TextWriterColor.WritePlain("", true);
            SuppressUploadMessage = false;
            if (IsError)
            {
                if (UploadNotificationProvoke)
                    UploadNotif.ProgressFailed = true;
                CancellationToken.Cancel();
                throw ReasonError;
            }
            else
            {
                if (UploadNotificationProvoke)
                    UploadNotif.Progress = 100;
                return true;
            }
        }

        /// <summary>
        /// Check for errors on download completion.
        /// </summary>
        private static void DownloadChecker(Exception e)
        {
            DebugWriter.Wdbg(DebugLevel.I, "Download complete. Error: {0}", e?.Message);
            if (e is not null)
            {
                if (DownloadNotificationProvoke)
                    DownloadNotif.ProgressFailed = true;
                ReasonError = e;
                IsError = true;
            }
            NetworkTools.TransferFinished = true;
        }

        /// <summary>
        /// Thread to check for errors on download completion.
        /// </summary>
        private static void UploadChecker(Exception e)
        {
            DebugWriter.Wdbg(DebugLevel.I, "Upload complete. Error: {0}", e?.Message);
            if (e is not null)
            {
                if (UploadNotificationProvoke)
                    UploadNotif.ProgressFailed = true;
                ReasonError = e;
                IsError = true;
            }
            NetworkTools.TransferFinished = true;
        }

        private static void HttpReceiveProgressWatch(object sender, HttpProgressEventArgs e)
        {
            int TotalBytes = (int)(e.TotalBytes ?? -1);
            var TransferInfo = new NetworkTransferInfo(e.BytesTransferred, TotalBytes, NetworkTransferType.Download);
            SuppressDownloadMessage = TotalBytes == -1;
            TransferProgress(TransferInfo);
        }

        private static void HttpSendProgressWatch(object sender, HttpProgressEventArgs e)
        {
            int TotalBytes = (int)(e.TotalBytes ?? -1);
            var TransferInfo = new NetworkTransferInfo(e.BytesTransferred, TotalBytes, NetworkTransferType.Upload);
            SuppressUploadMessage = TotalBytes == -1;
            TransferProgress(TransferInfo);
        }

        /// <summary>
        /// Report the progress to the console.
        /// </summary>
        private static void TransferProgress(NetworkTransferInfo TransferInfo)
        {
            try
            {
                // Distinguish download from upload
                bool NotificationProvoke = TransferInfo.TransferType == NetworkTransferType.Download ? DownloadNotificationProvoke : UploadNotificationProvoke;
                var NotificationInstance = TransferInfo.TransferType == NetworkTransferType.Download ? DownloadNotif : UploadNotif;

                // Report the progress
                if (!NetworkTools.TransferFinished)
                {
                    if (TransferInfo.FileSize >= 0L & !TransferInfo.MessageSuppressed)
                    {
                        // We know the total bytes. Print it out.
                        double Progress = 100d * (TransferInfo.DoneSize / (double)TransferInfo.FileSize);
                        if (NotificationProvoke)
                        {
                            NotificationInstance.Progress = (int)Math.Round(Progress);
                        }
                        else
                        {
                            if (!string.IsNullOrWhiteSpace(DownloadPercentagePrint))
                            {
                                TextWriterWhereColor.WriteWhere(PlaceParse.ProbePlaces(DownloadPercentagePrint), 0, ConsoleWrapper.CursorTop, false, KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Neutral), TransferInfo.DoneSize.FileSizeToString(), TransferInfo.FileSize.FileSizeToString(), Progress);
                            }
                            else
                            {
                                TextWriterWhereColor.WriteWhere(Translate.DoTranslation("{0} of {1} downloaded.") + " | {2}%", 0, ConsoleWrapper.CursorTop, false, KernelColorTools.GetConsoleColor(KernelColorTools.ColTypes.Neutral), TransferInfo.DoneSize.FileSizeToString(), TransferInfo.FileSize.FileSizeToString(), Progress);
                            }
                            ConsoleBase.ConsoleExtensions.ClearLineToRight();
                        }
                    }
                    else
                    {
                        TransferInfo.MessageSuppressed = true;
                    }
                }
            }
            catch (Exception ex)
            {
                DebugWriter.Wdbg(DebugLevel.E, "Error trying to report transfer progress: {0}", ex.Message);
                DebugWriter.WStkTrc(ex);
            }
        }

    }
}